File: FirebaseRemoteConfig/Sources/RCNUserDefaultsManager.m
/*
 * Copyright 2019 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FirebaseRemoteConfig/Sources/RCNUserDefaultsManager.h"
#import <FirebaseCore/FIRLogger.h>
#import <FirebaseRemoteConfig/FIRRemoteConfig.h>

static NSString *const kRCNGroupPrefix = @"group";
static NSString *const kRCNGroupSuffix = @"firebase";
static NSString *const kRCNUserDefaultsKeyNamelastETag = @"lastETag";
static NSString *const kRCNUserDefaultsKeyNamelastETagUpdateTime = @"lastETagUpdateTime";
static NSString *const kRCNUserDefaultsKeyNameLastSuccessfulFetchTime = @"lastSuccessfulFetchTime";
static NSString *const kRCNUserDefaultsKeyNamelastFetchStatus = @"lastFetchStatus";
static NSString *const kRCNUserDefaultsKeyNameIsClientThrottled =
    @"isClientThrottledWithExponentialBackoff";
static NSString *const kRCNUserDefaultsKeyNameThrottleEndTime = @"throttleEndTime";
static NSString *const kRCNUserDefaultsKeyNamecurrentThrottlingRetryInterval =
    @"currentThrottlingRetryInterval";

@interface RCNUserDefaultsManager () {
  /// User Defaults instance for this bundleID. NSUserDefaults is guaranteed to be thread-safe.
  NSUserDefaults *_userDefaults;
  /// The suite name for this user defaults instance. It is a combination of a prefix and the
  /// bundleID. This is because you cannot use just the bundleID of the current app as the suite
  /// name when initializing user defaults.
  NSString *_userDefaultsSuiteName;
  /// The FIRApp that this instance is scoped within.
  NSString *_firebaseAppName;
  /// The Firebase Namespace that this instance is scoped within.
  NSString *_firebaseNamespace;
  /// The bundleID of the app. In case of an extension, this will be the bundleID of the parent app.
  NSString *_bundleIdentifier;
}

@end

@implementation RCNUserDefaultsManager

#pragma mark Initializers.

/// Designated initializer.
- (instancetype)initWithAppName:(NSString *)appName
                       bundleID:(NSString *)bundleIdentifier
                      namespace:(NSString *)firebaseNamespace {
  self = [super init];
  if (self) {
    _firebaseAppName = appName;
    _bundleIdentifier = bundleIdentifier;
    NSInteger location = [firebaseNamespace rangeOfString:@":"].location;
    if (location == NSNotFound) {
      FIRLogError(kFIRLoggerRemoteConfig, @"I-RCN000064",
                  @"Error: Namespace %@ is not fully qualified app:namespace.", firebaseNamespace);
      _firebaseNamespace = firebaseNamespace;
    } else {
      _firebaseNamespace = [firebaseNamespace substringToIndex:location];
    }

    // Initialize the user defaults with a prefix and the bundleID. For app extensions, this will be
    // the bundleID of the app extension.
    _userDefaults =
        [RCNUserDefaultsManager sharedUserDefaultsForBundleIdentifier:_bundleIdentifier];
  }

  return self;
}

+ (NSUserDefaults *)sharedUserDefaultsForBundleIdentifier:(NSString *)bundleIdentifier {
  static dispatch_once_t onceToken;
  static NSUserDefaults *sharedInstance;
  dispatch_once(&onceToken, ^{
    NSString *userDefaultsSuiteName =
        [RCNUserDefaultsManager userDefaultsSuiteNameForBundleIdentifier:bundleIdentifier];
    sharedInstance = [[NSUserDefaults alloc] initWithSuiteName:userDefaultsSuiteName];
  });
  return sharedInstance;
}

+ (NSString *)userDefaultsSuiteNameForBundleIdentifier:(NSString *)bundleIdentifier {
  NSString *suiteName =
      [NSString stringWithFormat:@"%@.%@.%@", kRCNGroupPrefix, bundleIdentifier, kRCNGroupSuffix];
  return suiteName;
}

#pragma mark Public properties.

- (NSString *)lastETag {
  return [[self instanceUserDefaults] objectForKey:kRCNUserDefaultsKeyNamelastETag];
}

- (void)setLastETag:(NSString *)lastETag {
  if (lastETag) {
    [self setInstanceUserDefaultsValue:lastETag forKey:kRCNUserDefaultsKeyNamelastETag];
  }
}

- (NSTimeInterval)lastETagUpdateTime {
  NSNumber *lastETagUpdateTime =
      [[self instanceUserDefaults] objectForKey:kRCNUserDefaultsKeyNamelastETagUpdateTime];
  return lastETagUpdateTime.doubleValue;
}

- (void)setLastETagUpdateTime:(NSTimeInterval)lastETagUpdateTime {
  if (lastETagUpdateTime) {
    [self setInstanceUserDefaultsValue:@(lastETagUpdateTime)
                                forKey:kRCNUserDefaultsKeyNamelastETagUpdateTime];
  }
}

- (NSTimeInterval)lastFetchTime {
  NSNumber *lastFetchTime =
      [[self instanceUserDefaults] objectForKey:kRCNUserDefaultsKeyNameLastSuccessfulFetchTime];
  return lastFetchTime.doubleValue;
}

- (void)setLastFetchTime:(NSTimeInterval)lastFetchTime {
  [self setInstanceUserDefaultsValue:@(lastFetchTime)
                              forKey:kRCNUserDefaultsKeyNameLastSuccessfulFetchTime];
}

- (NSString *)lastFetchStatus {
  return [[self instanceUserDefaults] objectForKey:kRCNUserDefaultsKeyNamelastFetchStatus];
}

- (void)setLastFetchStatus:(NSString *)lastFetchStatus {
  if (lastFetchStatus) {
    [self setInstanceUserDefaultsValue:lastFetchStatus
                                forKey:kRCNUserDefaultsKeyNamelastFetchStatus];
  }
}

- (BOOL)isClientThrottledWithExponentialBackoff {
  NSNumber *isClientThrottled =
      [[self instanceUserDefaults] objectForKey:kRCNUserDefaultsKeyNameIsClientThrottled];
  return isClientThrottled.boolValue;
}

- (void)setIsClientThrottledWithExponentialBackoff:(BOOL)isClientThrottled {
  [self setInstanceUserDefaultsValue:@(isClientThrottled)
                              forKey:kRCNUserDefaultsKeyNameIsClientThrottled];
}

- (NSTimeInterval)throttleEndTime {
  NSNumber *throttleEndTime =
      [[self instanceUserDefaults] objectForKey:kRCNUserDefaultsKeyNameThrottleEndTime];
  return throttleEndTime.doubleValue;
}

- (void)setThrottleEndTime:(NSTimeInterval)throttleEndTime {
  [self setInstanceUserDefaultsValue:@(throttleEndTime)
                              forKey:kRCNUserDefaultsKeyNameThrottleEndTime];
}

- (NSTimeInterval)currentThrottlingRetryIntervalSeconds {
  NSNumber *throttleEndTime = [[self instanceUserDefaults]
      objectForKey:kRCNUserDefaultsKeyNamecurrentThrottlingRetryInterval];
  return throttleEndTime.doubleValue;
}

- (void)setCurrentThrottlingRetryIntervalSeconds:(NSTimeInterval)throttlingRetryIntervalSeconds {
  [self setInstanceUserDefaultsValue:@(throttlingRetryIntervalSeconds)
                              forKey:kRCNUserDefaultsKeyNamecurrentThrottlingRetryInterval];
}

#pragma mark Public methods.
- (void)resetUserDefaults {
  [self resetInstanceUserDefaults];
}

#pragma mark Private methods.

// There is a nested hierarchy for the userdefaults as follows:
// [FIRAppName][FIRNamespaceName][Key]
- (nonnull NSDictionary *)appUserDefaults {
  NSString *appPath = _firebaseAppName;
  NSDictionary *appDict = [_userDefaults valueForKeyPath:appPath];
  if (!appDict) {
    appDict = [[NSDictionary alloc] init];
  }
  return appDict;
}

// Search for the user defaults for this (app, namespace) instance using the valueForKeyPath method.
- (nonnull NSDictionary *)instanceUserDefaults {
  NSString *appNamespacePath =
      [NSString stringWithFormat:@"%@.%@", _firebaseAppName, _firebaseNamespace];
  NSDictionary *appNamespaceDict = [_userDefaults valueForKeyPath:appNamespacePath];

  if (!appNamespaceDict) {
    appNamespaceDict = [[NSMutableDictionary alloc] init];
  }
  return appNamespaceDict;
}

// Update users defaults for just this (app, namespace) instance.
- (void)setInstanceUserDefaultsValue:(NSObject *)value forKey:(NSString *)key {
  @synchronized(_userDefaults) {
    NSMutableDictionary *appUserDefaults = [[self appUserDefaults] mutableCopy];
    NSMutableDictionary *appNamespaceUserDefaults = [[self instanceUserDefaults] mutableCopy];
    [appNamespaceUserDefaults setObject:value forKey:key];
    [appUserDefaults setObject:appNamespaceUserDefaults forKey:_firebaseNamespace];
    [_userDefaults setObject:appUserDefaults forKey:_firebaseAppName];
    // We need to synchronize to have this value updated for the extension.
    [_userDefaults synchronize];
  }
}

// Delete any existing userdefaults for this instance.
- (void)resetInstanceUserDefaults {
  @synchronized(_userDefaults) {
    NSMutableDictionary *appUserDefaults = [[self appUserDefaults] mutableCopy];
    NSMutableDictionary *appNamespaceUserDefaults = [[self instanceUserDefaults] mutableCopy];
    [appNamespaceUserDefaults removeAllObjects];
    [appUserDefaults setObject:appNamespaceUserDefaults forKey:_firebaseNamespace];
    [_userDefaults setObject:appUserDefaults forKey:_firebaseAppName];
    // We need to synchronize to have this value updated for the extension.
    [_userDefaults synchronize];
  }
}

@end

File: FirebaseRemoteConfig/Sources/RCNUserDefaultsManager.swift
import Foundation
import FirebaseCore // For FIRLogger

// TODO: Move keys to a central constants file
private enum UserDefaultsKeys {
    static let lastETag = "lastETag"
    static let lastETagUpdateTime = "lastETagUpdateTime"
    static let lastSuccessfulFetchTime = "lastSuccessfulFetchTime"
    static let lastFetchStatus = "lastFetchStatus" // Note: This seems unused in RCNConfigSettingsInternal read path
    static let isClientThrottled = "isClientThrottledWithExponentialBackoff"
    static let throttleEndTime = "throttleEndTime"
    static let currentThrottlingRetryInterval = "currentThrottlingRetryInterval"
    static let realtimeThrottleEndTime = "throttleRealtimeEndTime"
    static let currentRealtimeThrottlingRetryInterval = "currentRealtimeThrottlingRetryInterval"
    static let realtimeRetryCount = "realtimeRetryCount"
    static let lastFetchedTemplateVersion = "fetchTemplateVersion" // From RCNConfigConstants? Check key name
    static let lastActiveTemplateVersion = "activeTemplateVersion" // From RCNConfigConstants? Check key name
    static let customSignals = "customSignals"

    // Grouping constants from ObjC implementation
    static let groupPrefix = "group"
    static let groupSuffix = "firebase"

}


/// Wraps UserDefaults to provide scoped, thread-safe access for Remote Config settings.
class RCNUserDefaultsManager {

    private let userDefaults: UserDefaults
    private let firebaseAppName: String
    private let firebaseNamespace: String // Just the namespace part (e.g., "firebase")
    private let bundleIdentifier: String
    private let lock = NSLock() // Lock for synchronizing writes

    // MARK: - Initialization

    /// Designated initializer.
    init(appName: String, bundleID: String, firebaseNamespace qualifiedNamespace: String) {
        self.firebaseAppName = appName
        self.bundleIdentifier = bundleID

        // Extract namespace part from "namespace:appName"
        if let range = qualifiedNamespace.range(of: ":") {
             self.firebaseNamespace = String(qualifiedNamespace[..<range.lowerBound])
        } else {
             // TODO: Log error - Namespace not fully qualified
             print("Error: Namespace '\(qualifiedNamespace)' is not fully qualified.")
             self.firebaseNamespace = qualifiedNamespace // Use as is, might cause issues
        }

        // Get shared UserDefaults instance for the app group derived from bundle ID
        self.userDefaults = RCNUserDefaultsManager.sharedUserDefaults(forBundleIdentifier: bundleID)
    }

    // MARK: - Static Methods for Shared UserDefaults

    private static func userDefaultsSuiteName(forBundleIdentifier bundleIdentifier: String) -> String {
        // Ensure bundleIdentifier is not empty? ObjC didn't check here.
        return "\(UserDefaultsKeys.groupPrefix).\(bundleIdentifier).\(UserDefaultsKeys.groupSuffix)"
    }

    private static func sharedUserDefaults(forBundleIdentifier bundleIdentifier: String) -> UserDefaults {
        // This mimics dispatch_once behavior implicitly through static initialization in Swift >= 1.2
        struct Static {
            static let instance = UserDefaults(suiteName: userDefaultsSuiteName(forBundleIdentifier: bundleIdentifier)) ?? UserDefaults.standard // Fallback unlikely needed
        }
        return Static.instance
    }

    // MARK: - Scoped Key Path

    /// Generates the key path for accessing the setting within UserDefaults: AppName.Namespace.Key
    private func scopedKeyPath(forKey key: String) -> String {
        return "\(firebaseAppName).\(firebaseNamespace).\(key)"
    }

    // MARK: - Read/Write Helpers (with locking)

    private func readValue<T>(forKey key: String) -> T? {
        // Reading UserDefaults is thread-safe, no lock needed technically,
        // but ObjC implementation read inside @synchronized block indirectly via instanceUserDefaults.
        // Let's keep reads simple for now.
        return userDefaults.value(forKeyPath: scopedKeyPath(forKey: key)) as? T
    }

    private func writeValue(_ value: Any?, forKey key: String) {
        lock.lock()
        defer { lock.unlock() }

        // We need to read the current app dictionary, then the namespace dictionary,
        // modify it, and write the whole app dictionary back. This mimics the ObjC logic.

        let appKey = firebaseAppName
        let namespaceKey = firebaseNamespace
        let settingKey = key

        var appDict = userDefaults.dictionary(forKey: appKey) ?? [:]
        var namespaceDict = appDict[namespaceKey] as? [String: Any] ?? [:]

        if let newValue = value {
            namespaceDict[settingKey] = newValue
        } else {
            namespaceDict.removeValue(forKey: settingKey) // Remove if value is nil
        }

        appDict[namespaceKey] = namespaceDict // Put potentially modified namespace dict back
        userDefaults.set(appDict, forKey: appKey) // Write the whole app dict back

        // Mimic explicit synchronize call, though often discouraged in Swift.
        // Required for potential app extension communication relying on it.
        userDefaults.synchronize()
    }


    // MARK: - Public Properties (Computed)

    var lastETag: String? {
        get { readValue(forKey: UserDefaultsKeys.lastETag) }
        set { writeValue(newValue, forKey: UserDefaultsKeys.lastETag) }
    }

    var lastETagUpdateTime: TimeInterval {
        get { readValue(forKey: UserDefaultsKeys.lastETagUpdateTime) ?? 0.0 }
        set { writeValue(newValue, forKey: UserDefaultsKeys.lastETagUpdateTime) }
    }

    var lastFetchTime: TimeInterval {
         get { readValue(forKey: UserDefaultsKeys.lastSuccessfulFetchTime) ?? 0.0 }
         set { writeValue(newValue, forKey: UserDefaultsKeys.lastSuccessfulFetchTime) }
     }

     // lastFetchStatus seems unused internally for read, only written? Keep setter for now.
     // var lastFetchStatus: String? {
     //     get { readValue(forKey: UserDefaultsKeys.lastFetchStatus) }
     //     set { writeValue(newValue, forKey: UserDefaultsKeys.lastFetchStatus) }
     // }

     var isClientThrottledWithExponentialBackoff: Bool {
         get { readValue(forKey: UserDefaultsKeys.isClientThrottled) ?? false }
         set { writeValue(newValue, forKey: UserDefaultsKeys.isClientThrottled) }
     }

     var throttleEndTime: TimeInterval {
         get { readValue(forKey: UserDefaultsKeys.throttleEndTime) ?? 0.0 }
         set { writeValue(newValue, forKey: UserDefaultsKeys.throttleEndTime) }
     }

     var currentThrottlingRetryIntervalSeconds: TimeInterval {
         get { readValue(forKey: UserDefaultsKeys.currentThrottlingRetryInterval) ?? 0.0 }
         set { writeValue(newValue, forKey: UserDefaultsKeys.currentThrottlingRetryInterval) }
     }

     var realtimeThrottleEndTime: TimeInterval {
         get { readValue(forKey: UserDefaultsKeys.realtimeThrottleEndTime) ?? 0.0 }
         set { writeValue(newValue, forKey: UserDefaultsKeys.realtimeThrottleEndTime) }
     }

     var currentRealtimeThrottlingRetryIntervalSeconds: TimeInterval {
          get { readValue(forKey: UserDefaultsKeys.currentRealtimeThrottlingRetryInterval) ?? 0.0 }
          set { writeValue(newValue, forKey: UserDefaultsKeys.currentRealtimeThrottlingRetryInterval) }
      }

      var realtimeRetryCount: Int {
           get { readValue(forKey: UserDefaultsKeys.realtimeRetryCount) ?? 0 }
           set { writeValue(newValue, forKey: UserDefaultsKeys.realtimeRetryCount) }
       }

     var lastFetchedTemplateVersion: String? { // Defaulted to "0" in ObjC getter if nil
         get { readValue(forKey: UserDefaultsKeys.lastFetchedTemplateVersion) ?? "0" }
         set { writeValue(newValue, forKey: UserDefaultsKeys.lastFetchedTemplateVersion) }
     }

     var lastActiveTemplateVersion: String? { // Defaulted to "0" in ObjC getter if nil
         get { readValue(forKey: UserDefaultsKeys.lastActiveTemplateVersion) ?? "0" }
         set { writeValue(newValue, forKey: UserDefaultsKeys.lastActiveTemplateVersion) }
     }

     var customSignals: [String: String] {
         get { readValue(forKey: UserDefaultsKeys.customSignals) ?? [:] } // Default to empty dict
         set { writeValue(newValue, forKey: UserDefaultsKeys.customSignals) }
     }


    // MARK: - Public Methods

    /// Delete all saved user defaults for this instance (App Name + Namespace scope).
    func resetUserDefaults() {
        lock.lock()
        defer { lock.unlock() }

        let appKey = firebaseAppName
        let namespaceKey = firebaseNamespace

        var appDict = userDefaults.dictionary(forKey: appKey) ?? [:]
        appDict.removeValue(forKey: namespaceKey) // Remove the namespace dict

        if appDict.isEmpty {
             userDefaults.removeObject(forKey: appKey) // Remove app dict if empty
        } else {
             userDefaults.set(appDict, forKey: appKey) // Write back modified app dict
        }

        userDefaults.synchronize()
    }

    // MARK: - Placeholder Selectors (for @objc calls from RCNConfigSettingsInternal)
    // These allow RCNConfigSettingsInternal to call this Swift class via selectors
    // until RCNConfigSettingsInternal is updated to call Swift methods directly.

    @objc func lastETagObjc() -> String? { return lastETag }
    @objc func setLastETagObjc(_ etag: String?) { lastETag = etag }

    @objc func lastETagUpdateTimeObjc() -> TimeInterval { return lastETagUpdateTime }
    @objc func setLastETagUpdateTimeObjc(_ time: TimeInterval) { lastETagUpdateTime = time }

    @objc func lastFetchTimeObjc() -> TimeInterval { return lastFetchTime }
    @objc func setLastFetchTimeObjc(_ time: TimeInterval) { lastFetchTime = time }

    // No getter for lastFetchStatus needed?
    // @objc func setLastFetchStatusObjc(_ status: String?) { lastFetchStatus = status }

    @objc func isClientThrottledWithExponentialBackoffObjc() -> Bool { return isClientThrottledWithExponentialBackoff }
    @objc func setIsClientThrottledWithExponentialBackoffObjc(_ throttled: Bool) { isClientThrottledWithExponentialBackoff = throttled }

    @objc func throttleEndTimeObjc() -> TimeInterval { return throttleEndTime }
    @objc func setThrottleEndTimeObjc(_ time: TimeInterval) { throttleEndTime = time }

    @objc func currentThrottlingRetryIntervalSecondsObjc() -> TimeInterval { return currentThrottlingRetryIntervalSeconds }
    @objc func setCurrentThrottlingRetryIntervalSecondsObjc(_ interval: TimeInterval) { currentThrottlingRetryIntervalSeconds = interval }

    @objc func realtimeThrottleEndTimeObjc() -> TimeInterval { return realtimeThrottleEndTime }
    @objc func setRealtimeThrottleEndTimeObjc(_ time: TimeInterval) { realtimeThrottleEndTime = time }

    @objc func currentRealtimeThrottlingRetryIntervalSecondsObjc() -> TimeInterval { return currentRealtimeThrottlingRetryIntervalSeconds }
    @objc func setCurrentRealtimeThrottlingRetryIntervalSecondsObjc(_ interval: TimeInterval) { currentRealtimeThrottlingRetryIntervalSeconds = interval }

    @objc func realtimeRetryCountObjc() -> Int { return realtimeRetryCount }
    @objc func setRealtimeRetryCountObjc(_ count: Int) { realtimeRetryCount = count }

    @objc func lastFetchedTemplateVersionObjc() -> String? { return lastFetchedTemplateVersion }
    @objc func setLastFetchedTemplateVersionObjc(_ version: String?) { lastFetchedTemplateVersion = version }

    @objc func lastActiveTemplateVersionObjc() -> String? { return lastActiveTemplateVersion }
    @objc func setLastActiveTemplateVersionObjc(_ version: String?) { lastActiveTemplateVersion = version }

    @objc func customSignalsObjc() -> [String: String]? { return customSignals }
    @objc func setCustomSignalsObjc(_ signals: [String: String]?) { customSignals = signals ?? [:] }

    @objc func resetUserDefaultsObjc() { resetUserDefaults() }

}

// Temporary placeholder for RemoteConfigSource enum if not defined elsewhere yet
//@objc(FIRRemoteConfigSource) public enum RemoteConfigSource: Int { case remote, defaultValue, staticValue }
